### PR TITLE
Implement the DUFNet

### PR DESCRIPTION
**Pull request by vic85821**

### CODE REVIEW

File: src/model/losses.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
+import torch
+import torch.nn as nn
+from torchvision import models
+
+
+class MISRL1Loss(nn.Module):
```
Comment (peter850706) [inline src/model/losses.py]: This class shoud be deleted.

File: src/model/losses.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    def forward(self, outputs, targets):
+        losses = [self.l1_loss(output, target) for output, target in zip(outputs, targets)]
+        loss = torch.stack(losses).mean()
+        return loss
+
+
+class HuberLoss(nn.Module):
+    """The implementation of the HuberLoss.
+    Args:
+        delta (float)
```
Comment (peter850706) [inline src/model/losses.py]: The comment is missing.

File: src/model/losses.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        quadratic = torch.min(abs_error, delta)
+        linear = (abs_error - quadratic)
+        losses = 0.5 * quadratic ** 2 + delta * linear
+        return torch.mean(losses)
+
+
+class PerceptualLoss(nn.Module):
+    """Use VGG16 as the feature extrator.
+    Args:
+        requires_grad (boolean)
```
Comment (peter850706) [inline src/model/losses.py]: The comment is missing.

File: src/model/nets/misr_duf_net.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        # The blocks in the end of the ResidualGenerationNetwork
+        self.residualNet = nn.Sequential()
+        self.residualNet.add_module('relu1', nn.ReLU())
+        self.residualNet.add_module('conv1', nn.Conv3d(256, 256, kernel_size=1))
+        self.residualNet.add_module('relu2', nn.ReLU())
+        self.residualNet.add_module('conv2', nn.Conv3d(256, in_channels * (upscale_factor ** 2), kernel_size=1))
+
+    def forward(self, inputs):
+        # dim: (B, C, 1, H, W)
+        target = inputs[self.num_frames // 2].unsqueeze(dim=2)
```
Comment (peter850706) [inline src/model/nets/misr_duf_net.py]: The same problem as described in AcdcMISRDUFDataset.

File: src/data/datasets/acdc_misr_duf_dataset.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            imgs = imgs[..., start:end]
+
+        hr_imgs = [imgs[..., t] for t in range(imgs.shape[-1])] # list of (H, W, C)
+
+        hr_imgs = self.transforms(*hr_imgs)
+        lr_imgs = self.degrade(*hr_imgs)
+        lr_imgs = self.post_transforms(*lr_imgs)
+        hr_imgs = self.post_transforms(*hr_imgs)
+        lr_imgs = [img.permute(2, 0, 1).contiguous() for img in lr_imgs]
+        hr_img = hr_imgs[self.num_frames // 2].permute(2, 0, 1).contiguous()
```
[the remark this answers is not in the document]
Comment (vic85821) [reply src/data/datasets/acdc_misr_duf_dataset.py]: Solved.

File: src/model/losses.py (hunk 1 above)
Comment (vic85821) [reply src/model/losses.py]: Deleted.

**Pull request merged**